src/python/pants/build_graph/build_configuration.py

```diff
@@ -22,6 +22,9 @@
 class BuildConfiguration:
   """Stores the types and helper functions exposed to BUILD files."""
 
+  def __str__(self):
+    return 'BuildConfiguration(rules=[set of size {}])'.format(len(self._rules))
```

Comment:
Consider mentioning the other members, e.g. union rules and targets.

I wish this were a dataclass so that we got this for free. Maybe it would be possible (in a followup) to make that refactor.